apollo-federation/src/operation/mod.rs

```diff
@@ -3922,6 +3648,400 @@ impl NamedFragments {
+                inline_fragment_selection.has_defer()
+            }
+        }
+    }
+
+    fn without_defer(
+        &self,
+        filter: DeferFilter<'_>,
+        named_fragments: &NamedFragments,
+    ) -> Result<Self, FederationError> {
+        match self {
+            Selection::Field(field) => {
+                let Some(selection_set) = field
+                    .selection_set
+                    .as_ref()
+                    .filter(|selection_set| selection_set.has_defer())
+                else {
+                    return Ok(Selection::Field(Arc::clone(field)));
+                };
+
+                Ok(field
+                    .with_updated_selection_set(Some(
+                        selection_set.without_defer(filter, named_fragments)?,
+                    ))
+                    .into())
+            }
+            Selection::FragmentSpread(frag) => {
+                let spread = frag.spread.without_defer(filter)?;
+                Ok(FragmentSpreadSelection::new(spread, named_fragments)?.into())
+            }
+            Selection::InlineFragment(frag) => {
+                let inline_fragment = frag.inline_fragment.without_defer(filter)?;
+                let selection_set = frag.selection_set.without_defer(filter, named_fragments)?;
+                Ok(InlineFragmentSelection::new(inline_fragment, selection_set).into())
+            }
+        }
+    }
+
+    fn normalize_defer(self, normalizer: &mut DeferNormalizer) -> Result<Self, FederationError> {
+        match self {
+            Selection::Field(field) => Ok(Self::Field(Arc::new(
+                field.with_updated_selection_set(
+                    field
+                        .selection_set
+                        .clone()
+                        .map(|set| set.normalize_defer(normalizer))
+                        .transpose()?,
+                ),
+            ))),
+            Selection::FragmentSpread(_spread) => {
+                Err(FederationError::internal("unexpected fragment spread"))
+            }
+            Selection::InlineFragment(inline) => inline
+                .with_updated_selection_set(
+                    inline.selection_set.clone().normalize_defer(normalizer)?,
+                )
+                .normalize_defer(normalizer)
+                .map(|inline| Self::InlineFragment(Arc::new(inline))),
+        }
+    }
+}
+
+impl SelectionSet {
+    /// Create a new selection set without @defer directive applications.
+    fn without_defer(
+        &self,
+        filter: DeferFilter<'_>,
+        named_fragments: &NamedFragments,
+    ) -> Result<Self, FederationError> {
+        let mut without_defer =
+            SelectionSet::empty(self.schema.clone(), self.type_position.clone());
+        for selection in self.selections.values() {
+            without_defer
+                .add_local_selection(&selection.without_defer(filter, named_fragments)?, true)?;
+        }
+        debug_assert!(!without_defer.has_defer());
```

Comment:
This is only true if `filter` is `All`.

Reply:
A! good spot, this check should only happen in `Operation::without_defer`